src/modules/logger/log_writer_file.cpp

```diff
@@ -81,6 +82,111 @@ LogWriterFile::~LogWriterFile()
 	pthread_cond_destroy(&_cv);
 }
 
+#if defined(PX4_CRYPTO)
+bool LogWriterFile::init_logfile_encryption(const char *filename)
+{
+	if (_algorithm == CRYPTO_NONE) {
+		return true;
+	}
+
+	// open a crypto session // TODO: use a parameter for algorithm
+	if (!_crypto.open(_algorithm)) {
+		return false;
+	}
+
+	// Get the minimum block size for which the encryption can be performed
+	_min_blocksize = _crypto.get_min_blocksize(_key_idx);
+
+	// Generate a crypto key and store it to the keystore.
+
+	// TODO: use parameter for alogrithm? Or product specific macro?
+	// TODO: how to define key index to the keystore? Product macro? parameter?
+	if (!_crypto.generate_key(_key_idx, false)) {
+		PX4_ERR("Can't generate crypto key");
+		return false;
+	}
+
+	// Get the generated key, encrypted with RSA_OAEP. Open another temporary session for this.
+	PX4Crypto rsa_crypto;
+
+	if (!rsa_crypto.open(CRYPTO_RSA_OAEP)) {
+		return false;
+	}
+
+	/* Get the size of an encrypted key and nonce */
+
+	size_t key_size;
+	size_t nonce_size;
+
+	if (!rsa_crypto.get_encrypted_key(_key_idx,
+					  NULL,
+					  &key_size,
+					  _exchange_key_idx) ||
+	    !_crypto.get_nonce(NULL, &nonce_size) ||
+	    key_size == 0) {
+		rsa_crypto.close();
+		return false;
+	}
+
+	/* Allocate space and get key + nonce */
+	uint8_t *key = (uint8_t *)malloc(key_size + nonce_size);
+
+	if (!key ||
+	    !rsa_crypto.get_encrypted_key(
+		    _key_idx,
+		    key,
+		    &key_size,
+		    _exchange_key_idx) ||
+	    !_crypto.get_nonce(
+		    key + key_size, &nonce_size)) {
+		PX4_ERR("Can't get & encrypt the key");
+		free(key);
+		rsa_crypto.close();
+		return false;
+	}
+
+	rsa_crypto.close();
+
+	// Write the encrypted key to the disk
+
+	size_t fnlen = strlen(filename);
+	char *tmp_buf = (char *)malloc(fnlen + 1);
+
+	if (!tmp_buf) {
+		PX4_ERR("out of memory");
```

Comment:
I think `free(key)` is missing for that case.

Reply:
correct, thanks! Will fix this